ticker/src/main/java/com/robinhood/ticker/TickerCharacterList.java

```diff
@@ -0,0 +1,114 @@
+ * This is the primary class that Ticker uses to determine how to animate from one character
+ * to another. The provided string dictates what characters will appear between
+ * the start and end characters.
+ *
+ * <p>For example, given the string "abcde", if the view wants to animate from 'd' to 'b',
+ * it will know that it has to go from 'd' to 'c' to 'b', and these are the characters
+ * that show up during the animation scroll.
+ *
+ * @author Jin Cao, Robinhood
+ */
+public class TickerCharacterList {
+    private final int numOriginalCharacters;
+    // The saved character list will always be of the format: EMPTY, list, list
+    private final char[] characterList;
+    // A minor optimization so that we can cache the indices of each character.
+    private final Map<Character, Integer> characterIndicesMap;
+
+    public TickerCharacterList(String characterList) {
+        if (characterList.contains(Character.toString(TickerUtils.EMPTY_CHAR))) {
+            throw new IllegalArgumentException(
+                    "You don't need to include TickerUtils.EMPTY_CHAR in the character list.");
+        }
+
+        final char[] charsArray = characterList.toCharArray();
+        final int length = charsArray.length;
+        this.numOriginalCharacters = length;
+
+        characterIndicesMap = new HashMap<>(length);
+        for (int i = 0; i < length; i++) {
+            characterIndicesMap.put(charsArray[i], i);
+        }
+
+        this.characterList = new char[length * 2 + 1];
+        this.characterList[0] = TickerUtils.EMPTY_CHAR;
+        for (int i = 0; i < length; i++) {
+            this.characterList[1 + i] = charsArray[i];
+            this.characterList[1 + length + i] = charsArray[i];
+        }
+    }
+
+    /**
+     * @param start the character that we want to animate from
+     * @param end the character that we want to animate to
+     * @return a valid {@link AnimationCharacterIndices}, or null if the inputs are not supported.
+     */
+    AnimationCharacterIndices getCharacterIndices(char start, char end) {
+        int startIndex = getIndexOfChar(start);
+        int endIndex = getIndexOfChar(end);
+        if (startIndex < 0 || endIndex < 0) {
+            return null;
+        }
+
+        if (start != TickerUtils.EMPTY_CHAR && end != TickerUtils.EMPTY_CHAR &&
+                endIndex < startIndex) {
+            final int nonWrapDistance = startIndex - endIndex;
+            final int wrapDistance = numOriginalCharacters - startIndex + endIndex;
+            if (wrapDistance < nonWrapDistance) {
+                endIndex = numOriginalCharacters + endIndex;
+            }
+        }
+        return new AnimationCharacterIndices(startIndex, endIndex);
+    }
+
+    Set<Character> getSupportedCharacters() {
+       return characterIndicesMap.keySet();
+    }
+
+    char[] getCharacterList() {
+        return characterList;
+    }
+
+    private int getIndexOfChar(char c) {
+        if (c == TickerUtils.EMPTY_CHAR) {
```

Comment:
i'm not loving the special case handling of the empty_char. is there any way to get this for free? like enforcing the empty_char is always at 0? or why do we even need it at 0? can we eliminate that restriction?

Reply:
looping back here - maybe user passes in "abcdef...", and when we wrap to the TickerCharacterList, we add the empty_char at index 0, then don't have to special case here?

Reply:
so i went back and forth on this one for a while. the reason why the empty char is needed in the saved character list is that we use the character list for animation purposes (so we don't have to construct a new char array for each animation). I originally played around with saving the empty char index (which is 0) into the map. however, this actually made wrap-around animation a little more complicated than i would like.

e.g., given the characters "abcd", the persisted character list for animation is "_abcdabcd". the reason why we save two copies is so that we can easily animate from 'd' -> 'a' this way. Now, if we persist the actual indices, _ would be 0, 'a' would be 1, 'b' would be 2, and so on. However, computing the target distance wraparound would then become something like `numOriginalCharacters - startIndex - 1 + endIndex`, and I didn't like the arbitrary 1 there. 

i liked this way slightly more in the end because from the perspective of character indices there is no concept of added characters, and let `getIndexOfChar` deal with manipulating the saved indices.

eh, i can go either way at this point. im gonna see if i can think of a better way to handle empty chars in general. the main pain point as pointed out earlier is that i want to have it included in the main char array for animation purposes which i dont think we can avoid.